### DISCUSSION TITLE
Problem with Script

### DISCUSSION THREAD

**Discussion by puskyer** [arendst/Tasmota#16272] [Q&A] [answered]
hello, created my first script on an ESP-01 (1M flash) DHT-11 module, I want to use to monitor humidity in a location.

part of the user_config_override.h for this Standard Tasmota build
`#define doemail

  #ifndef USE_SCRIPT
    #define USE_SCRIPT // adds about 17k flash size, variable ram size
  #endif
  
  #ifdef USE_RULES
    #undef USE_RULES
  #endif

  #define DEBUG_EMAIL_PORT Serial
  #define USE_BUTTON_EVENT
  #define SCRIPT_POWER_SECTION
  #define USE_SCRIPT_WEB_DISPLAY
  #define SUPPORT_MQTT_EVENT
  #define SUPPORT_IF_STATEMENT
  #define USE_SCRIPT_SUB_COMMAND

  #ifdef doemail
    //#define USE_TELEGRAM
    #ifdef ESP8266
    #define USE_SENDMAIL
    #endif
    #ifdef ESP32
    #undef USE_SENDMAI
    #define USE_ESP32MAIL
    #endif
    #define EMAIL_USER "email address"
    #define EMAIL_PASSWORD "app password"
    #ifdef USE_ESP32MAIL
       #undef EMAIL_FROM
       #define EMAIL_FROM "email address"
    #else
       #define EMAIL_FROM "<email address>"
    #endif
    #define EMAIL_SERVER "smtp.gmail.com"
    #define EMAIL_PORT 465
    #define MAIL_TIMEOUT 2000
  #endif
`

MQTT and emulation disabled

Program Version	12.1.0.1(tasmota)
Build Date & Time	2022-08-18T12:52:16
Core/SDK Version	2_7_4_9/2.2.2-dev(38a443e)
ESP Chip Id	7498020 (ESP8266EX)
Flash Chip Id	0x1440D8
Flash Size	1024 KB
Program Flash Size	1024 KB
Program Size	711 KB
Free Program Space	292 KB
Free Memory	23.6 KB 

The script that is not working is the following
`>D 25
p:hh=75
p:lh=40

hr=0
hm=0
tp=0
dp=0
tpd="C"
pr=0
email="email@outlook.com"

>T
hm=DHT11#Humidity
tp=DHT11#Temperature
dp=DHT11#DewPoint
tpd=DHT11#TempUnit

; if humidity is > 75 or < 40 send email once
if ((hm>=hh or hm<=lh) and pr==0)
then
pr=1
print  sending email
->sendmail [*:*:*:*:*:<%email%>:Humidity Alarm]*
endif

if (hm<=hh or hm>=hl)
then
print check
pr=0
endif

print  paramiters are  %pr% %hm% %hh% %lh%
; send every hour
hr=hours
if chg[hr]>0
and pr==1
then
print  sending repet email
;->sendmail [*:*:*:*:*:<%email%>: Humidity
Alarm repeted]*
endif

>m
Date %tstamp%
Humidity is %1hm% %%
Temperature is %1tp% °%tpd%
DewPoint is %1dp% °%tpd%
#`

 console output

`14:58:54.880 sending email
14:59:01.745 check
14:59:01.748 paramiters are  0.00 39.00 0.00 0.00
14:59:02.478 sending email
14:59:13.831 check
14:59:13.834 paramiters are  0.00 39.00 0.00 0.00
14:59:14.420 sending email
14:59:21.759 check
14:59:21.762 paramiters are  0.00 39.00 0.00 0.00
14:59:22.375 sending email
14:59:29.909 check
14:59:29.912 paramiters are  0.00 38.00 0.00 0.00`

First Question , so why are my p:hh=75 and p:lh=40 not printing correctly?
Second it should send the email once when the humidity is > 75 or < 40 and then once an hour after that... yet the email gets sent approximately every 10 seconds??

**Comment by barbudor** [comment, marked as answer]
2nd question:
This test is wrong : `if (hm<=hh or hm>=hl)`
It should be an `and` as well as using strict `<` and `>` 

no idea for the 1st question, I'm not a script guru

**Reply by puskyer** [reply]
thank you for the reply, i will try it with your recommendation..

**Reply by puskyer** [reply]
tried the or and it stopped sending the email, but looked at  the logic again and I believe an "or" is required.

ie  I want an email to be sent if the humidity  goes below 40% or if the humidity  goes higher then 75%

if I do an "and" then humidity can never be <40% and >75% at the same time?? does my logic make sense?

**Reply by puskyer** [reply]
Sorry was looking at the wrong code  I changed the if (hm<hh or hm>hl) to if (hm<hh and hm>hl)

thank you

**Reply by barbudor** [reply]
The condition for sending email is
Hum <= HumLow OR Hum >= HumHigh (outside or equal to low/high) 
The logical opposition condition is
Hum > HumLow AND Hum < HumHigh (strictly inside low/high)

With an OR a value that check test1 would also check test2
Try with 10 for example
Test1 would be true because 10<40
Test2 would be true because 10<75

**Reply by puskyer** [reply]
Thank you seems to be working now..
Will let it run for a few days..

**Comment by puskyer** [comment]
thank you @barbudor  on the assistance on my logic clarification your assistance was appreciated.

also found that the p: variables were not working as stated in the documentation, so I changed my minhm and maxhm to regular variables and all is working much better, will open a separate discussion on the p: variables...
